design/ingressroute-design.md

```diff
@@ -0,0 +1,300 @@
+
+Each _root_ of a DAG starts at a virtual host, which describes properties like the fully qualified name of the virtual host, any aliases (for example a www. prefix) of that vhost, TLS configuration, and possibly global access list details.
+The edges of a graph do not contain virtual host information, they are only reachable from a root via a delegation.
+This permits the _owner_ of an ingress root to both delegate the authority to publish a service on a portion of the route space inside a virtual host, and to itself further delegate authority to publish and delegate.
+
+This also means that 
+
+In practice the linkage, or delegation, from root to edge, is performed with a specific type of route action.
+You can think of it that rather than routing traffic to a service, it is routed to another ingressroute object for further processing.
+
+
+##
+
+# Detailed Design
+
+## IngressRoute CRD
+
+This is an example of a fully populated root ingressroute.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google
+  namespace: prod
+spec:
+  # virtualhost appears at most once. If it is present, the object is considered
+  # to be a "root".
+  virtualhost:
+    # the fully qualified domain name of the root of the ingress tree
+    # all leaves of the DAG rooted at this object relate to the fqdn (and its aliases)
+    fqdn: www.google.com
+    # a set of aliases for the domain, these may be alternative fqdns which are considered
+    # aliases of the primary fqdn
+    aliases:
+      - www.google.com.au
+      - google.com
+    # if present describes tls properties. The CNI names that will be matched on
+    # are described in fqdn and aliases, the tls.secretName secret must contain a
+    # matching certificate
+    tls:
+      # required, the name of a secret in the current namespace
+      secretName: google-tls
+      # other properties like cipher suites may be added later
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific.
+  routes:
+  # each route entry starts with a prefix match
+  # and one of service or delegate
+  - match: /static
+    # service defines the properties of the service in the current namespace
+    # that will handle traffic matching the route
+    service:
+    - name: google-static
+      port: 9000
+  - match: /finance
+    # delegate delegates the matching route to another IngressRoute object.
+    # This delegates the responsibility for /finance to the IngressRoute matching the delegate parameters
+    delegate:
+      # the name of the IngressRoute object to delegate to
+      name: google-finance
+      # the namespace of the IngressRoute object, if blank the current namespace is assumed
+      namespace: finance
```

Comment:
Good question. There is also another problem that if I know the name and namespace of your ingressroute object, I can construct a root or intermediate delegate ingress route and expose your website's routes on a domain of my control. I've added a section to the security concerns section.

Reply:
@stevesloka the ability for the author of an ingresroute crd in namespace X to know _which_ object they are trying to match for delegation is important. In the security section (I haven't gotten there yet -- TODO(dfc) update this comment) I mentioned that we might want a two way association and this would give the author of the delegate document a little more information on what they are delegated from.